### PR TITLE
Stop dead client connection from blocking instance snapshot / remove 

### LINKED ISSUE #12611: stalled remote client prevents snapshots from taking place
stalled remote client doing lxc pull (i.e. forking a lxc forkfile process) can block snapshot creation/deletion

see https://discourse.ubuntu.com/t/lxc-snapshot-operations-blocked-by-hanging-forkfile-process/40817 for discussion

how to reproduce:

on the client:
lxc file pull container/some/big/file

forcefully cut the network connection, i.e. pull the network cable, turn off the vpn, firewall off port 8443 etc

lxc snapshot container will hang
lxc rm container/snap0 will also hang

forkfile process spawned off of a lxd will also hang on the server indefinitely

### PR DESCRIPTION
**Pull request by gabrielmougard**
closes #12611  

___

@tomponline @roosterfish Here is an example of a reproducer with the patch (left is the server machine, right is the client machine. At some point, I create a network outage in the client side. The server is now able to finish its snapshotting operation even when the client died pulling for some files):

https://github.com/canonical/lxd/assets/27967361/3c8b3e21-fb86-43c8-928d-7e0958dd50d6

### CODE REVIEW

**Comment by gabrielmougard** [comment]
Integration tests WIP

File: lxd/response/response.go [hunk 1]
```diff
@@ -462,12 +553,17 @@ func (r *fileResponse) Render(w http.ResponseWriter) error {
 			rd = fd
 		}
 
+		// we apply a similar strategy as the single file case by wrapping
+		// each part's reader with a TimeoutReader. This allows us to manage TCP connection
+		// deadlines based on read activity for each part of the multipart response.
+		tr := NewTimeoutReader(rd, tcpConn, 1*time.Minute)
```
Comment (roosterfish) [inline lxd/response/response.go:559]: Can't this be a `NewTimeoutReadSeeker()` too? Then you could get rid of `NewTimeoutReader()`.

**Comment by tomponline** [comment]
@gabrielmougard as discussed in our 1:1 I think using the existing `SetTimeouts()` function like we do here:

https://github.com/canonical/lxd/blob/8f920ac525f986566c233db3d171e955bd306a38/lxd/instance_sftp.go#L126-L133

Will be a good way to fix this by using underlying TCP keepalives.

Also please do test for a reproducer before and then apply this fix and see if it helps.

Thanks

**Comment by gabrielmougard** [comment]
@tomponline it should be ready.

**Comment by tomponline** [comment]
@gabrielmougard as discussed in our 1:1 lets try using `conn := ucred.GetConnFromContext(r.Context())` instead of `Hijack()` to get underlying conn as this should make things more idiomatic and simpler.

**Comment by gabrielmougard** [comment]
It works fine with the `ucred.GetConnFromContext(r.Context())` approach. Thanks @tomponline for the tip!

File: lxd/response/response.go [hunk 2]
```diff
@@ -401,6 +403,19 @@ func (r *fileResponse) Render(w http.ResponseWriter) error {
 
 	// For a single file, return it inline
 	if len(r.files) == 1 {
+		remoteConn := ucred.GetConnFromContext(r.req.Context())
+
+		defer func() { _ = remoteConn.Close() }()
```
Comment (tomponline) [inline lxd/response/response.go:408]: This shouldnt be needed as Go's HTTP handler should take care of that.

File: lxd/response/response.go [hunk 3]
```diff
@@ -436,10 +451,15 @@ func (r *fileResponse) Render(w http.ResponseWriter) error {
 		w.Header().Set("Content-Disposition", fmt.Sprintf("inline;filename=%s", r.files[0].Filename))
 
 		http.ServeContent(w, r.req, r.files[0].Filename, mt, rs)
-
 		return nil
 	}
 
+	if r.headers != nil {
```
Comment (tomponline) [inline lxd/response/response.go:457]: What is this bit for?

Comment (gabrielmougard) [reply lxd/response/response.go:457]: Ah ! something I forget to remove !

**Review by tomponline — APPROVED** [review]

**Pull request merged**